### DISCUSSION TITLE
mixins integration with vscode-extension

### DISCUSSION THREAD

**Discussion by MChorfa** [getporter/porter#1327] [Q&A] [answered]
What are the steps that a mixin author have to do, so the vscode-extension can interpret the mixin schema ?
Related issue: MChorfa/porter-helm3#10

**Comment by carolynvs** [comment, marked as answer]
You will want to implement the [schema](https://porter.sh/mixin-dev-guide/commands/#schema) command for your mixin. When the VS Code extension loads, it runs `porter schema` which in turn calls `MIXIN schema` against every mixin and composes a single schema file that it provides to VS Code.

Your mixin is responsible for defining the allowed schema in jsonschema **for each action** because for some mixins the schema is different per action. Porter will handle parsing through your schema and piecing it all together into a coherent schema for porter.yaml. You just need to worry about what your schema should look like for your portion of the manifest related to the mixin:

```yaml
install:
- helm:
    description: Install my wonderful chart
    name: myrelease
    chart: ./charts/mychart
    ...
```

The way all the mixins implement that command is by having a packr template schema.json file built into their binary that they print out for the `schema` command.

Here is the schema.json for the helm2 mixin: https://github.com/getporter/helm-mixin/blob/main/pkg/helm/schema/schema.json. I recommend copying skeletors new implementation for the schema command and tests. Here's what the [test should look like](https://github.com/getporter/skeletor/blob/main/pkg/skeletor/schema_test.go) and also the [integration test](https://github.com/getporter/skeletor/blob/main/tests/schema_integration_test.go). 

The skeletor makefile [changed](https://github.com/getporter/skeletor/blob/6340bfb41c435d1b06252515af935e55b03b9735/Makefile#L73-L76) to add the integration testing, so you may need to sync back up with it to get that working again. The integration test is useful because otherwise you may see the schema working on your local machine but not in the cross compiled binaries that you publish which is a pain to troubleshoot. This catches that before you release.

Really the hard part is getting the json schema right. I recommend for that using https://www.jsonschemavalidator.net/ and https://www.convertjson.com/yaml-to-json.htm to validate that your schema is correct. Also unit tests around that help a lot too! But while you are playing around with it, those web pages are very useful and give you instant feedback and tell you line numbers and where you messed up the json schema. Which ... is where I usually need the most help! 😅

**Reply by MChorfa** [reply]
Thank you for this thorough explanation. It helped a lot 😄
